internal/controller/services/gateway/ocp_routes.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2025.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gateway
+
+// +kubebuilder:rbac:groups=route.openshift.io,resources=routes/custom-host,verbs=create;patch
+
+import (
+	"context"
+	"fmt"
+
+	routev1 "github.com/openshift/api/route/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	serviceApi "github.com/opendatahub-io/opendatahub-operator/v2/api/services/v1alpha1"
+	odhtypes "github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/metadata/labels"
+)
+
+// GatewayServiceFullName is the name of the auto-created Gateway service.
+// Format: <gateway-name>-<gatewayclass-name>.
+var GatewayServiceFullName = DefaultGatewayName + "-" + GatewayClassName
+
+// httpRouteReferencesGateway checks if an HTTPRoute references our gateway.
+func httpRouteReferencesGateway(httpRoute *gwapiv1.HTTPRoute) bool {
+	for _, ref := range httpRoute.Spec.ParentRefs {
+		refNamespace := GatewayNamespace
+		if ref.Namespace != nil {
+			refNamespace = string(*ref.Namespace)
+		}
+		if string(ref.Name) == DefaultGatewayName && refNamespace == GatewayNamespace {
+			return true
+		}
+	}
+	return false
+}
+
+// HTTPRouteGatewayRefPredicate filters HTTPRoutes that reference our gateway.
+func HTTPRouteGatewayRefPredicate() predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			httpRoute, ok := e.Object.(*gwapiv1.HTTPRoute)
+			if !ok {
+				return false
+			}
+			return httpRouteReferencesGateway(httpRoute)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			httpRoute, ok := e.ObjectNew.(*gwapiv1.HTTPRoute)
+			if !ok {
+				return false
+			}
+			return httpRouteReferencesGateway(httpRoute)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			httpRoute, ok := e.Object.(*gwapiv1.HTTPRoute)
+			if !ok {
+				return false
+			}
+			return httpRouteReferencesGateway(httpRoute)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			httpRoute, ok := e.Object.(*gwapiv1.HTTPRoute)
+			if !ok {
+				return false
+			}
+			return httpRouteReferencesGateway(httpRoute)
+		},
+	}
+}
```

Comment:
move this predicate function to pkg/controller/predicates/resources/resources.go